### PR TITLE
Feature/admin

### PR DESCRIPTION
**Pull request by dacur**
## Why?

The bookstore needs an admin feature
## What Changed?

Set up active admin
Added steps/tests, which are all passing
Cleaned up Book model and controller

### CODE REVIEW

File: app/admin/dashboard.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    #     end
+    #   end
+
+    #   column do
+    #     panel "Info" do
+    #       para "Welcome to ActiveAdmin."
+    #     end
+    #   end
+    # end
+  end # content
```
Comment (dkniffin) [inline app/admin/dashboard.rb]: You can delete these template comments

File: app/assets/stylesheets/active_admin.scss [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// SASS variable overrides must be declared before loading up Active Admin's styles.
+//
+// To view the variables that Active Admin provides, take a look at
+// `app/assets/stylesheets/active_admin/mixins/_variables.css.scss` in the
+// Active Admin source.
+//
+// For example, to change the sidebar width:
+// $sidebar-width: 242px;
+
+// Active Admin's got SASS!
```
Comment (dkniffin) [inline app/assets/stylesheets/active_admin.scss]: These comments can be deleted as well

File: app/models/admin_user.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  devise :database_authenticatable, 
+         :recoverable, :rememberable, :trackable, :validatable
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, 
+         :recoverable, :rememberable, :trackable, :validatable
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, 
+         :recoverable, :rememberable, :trackable, :validatable
```
Comment (dkniffin) [inline app/models/admin_user.rb]: It looks like when we ran the `active_admin:install` command, it inserted this three times. It should only be in here once

File: app/models/book.rb [hunk 4]
```diff
@@ -1,6 +1,5 @@
 class Book < ActiveRecord::Base
-  # default_scope { order(published_date: :desc) }
-  self.per_page = 25
+  # self.per_page = 25
```
Comment (dkniffin) [inline app/models/book.rb]: To quote Zach, "Commented code is evil!"

File: features/administration.feature [hunk 5]
```diff
@@ -48,12 +51,13 @@ Feature: Bookstore Administration Panel
...
     When I visit the admin books url
-    And I click "Edit" for the book "Book To Be Edited"
+    And I edit the book with the title "Book To Be Edited"
     And I change the book name to "Book That Has Been Edited"
     And I change the book price to "222.22"
-    And I click "Submit"
+    And I click the "Update Book" button
     And I visit the admin books url
     Then I see the book "Book That Has Been Edited"
     When I visit the public book index
```
Comment (dkniffin) [inline features/administration.feature:60]: These changes might be okay, but I think the idea of this exercise is to write code that meets the tests, not change the tests. @iandonovan should chime in on this one.

File: features/step_definitions/admin_steps.rb [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  fill_in "Email*", with: "admin@example.com"
+end
+
+When(/^I enter my admin password$/) do
+  fill_in "Password*", with: "password"
+end
+
+When(/^I click submit$/) do
+  click_button("Login")
+end
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb]: This is misleading. The test doesn't say "I click Login", it says "I click submit". Either you should change your test to say `When I click "Login"`, or you should change your button text to say "submit".

File: features/administration.feature (hunk 5 above)
Comment (dacur) [inline features/administration.feature:60]: The button is an Update Book button, not submit, so I changed it to match what the button says.  Should I change the button to Submit, or is there a test for just standard submit buttons?  Thanks

File: features/step_definitions/admin_steps.rb [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+Then(/^I see the admin panel$/) do
+  expect(page).to have_content("Dashboard")
+end
+
+Given(/^I do not have an admin account$/) do
+  AdminUser.delete_all
+end
+
+Then(/^I see a flash notification that tell me that my email does not exist in the system$/) do
+  expect(page).to have_content("Invalid email or password")
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb]: I think this should be more specific. You're simply looking for the text. You can also key on the html class, to make sure it's actually in a flash notification.

File: features/step_definitions/admin_steps.rb [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+end
+
+Given(/^I am logged into the admin panel$/) do
+  @admin = AdminUser.create(email: "admin@example.com", password: "password", 
+    password_confirmation: "password")
+    visit("/admin")
+    fill_in "Email*", with: "admin@example.com"
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb]: Not 100% sure, but you might be able to drop the "*". it depends on how that's implemented.

File: features/step_definitions/admin_steps.rb [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    fill_in "Password*", with: "password"
+    click_button("Login")
+end
+
+When(/^I visit the admin books url$/) do
+  visit("/admin/books")
+end
+
+When(/^I want to add a book$/) do
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb]: "When I visit the admin new books url" sounds better.

File: features/step_definitions/admin_steps.rb (hunk 6 above)
Comment (dacur) [inline features/step_definitions/admin_steps.rb]: Thanks, it turned out I didn't need this step, as it's covered elsewhere.

File: features/step_definitions/admin_steps.rb [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  visit("/admin/books/new")
+end
+
+When(/^I enter the title "(.*?)"$/) do |title|
+  fill_in "Title", with: title
+end
+
+When(/^I enter the price "(.*?)"$/) do |price|
+  fill_in "Price cents", with: price
+end
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb]: See if you can DRY up this and the previous `When`, using regex. Let me know if you need some help.

File: features/step_definitions/admin_steps.rb [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+  select year, from: "Year"
+end
+
+When(/^I select month "(.*?)"$/) do |month|
+  select month, from: "Month"
+end
+
+When(/^I select day "(.*?)"$/) do |day|
+  select day, from: "Day"
+end
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb:61]: These date selections are fine, but an easier way might be to have a single `I select the date....` step.

File: features/step_definitions/admin_steps.rb (hunk 9 above)
Comment (dacur) [inline features/step_definitions/admin_steps.rb]: I agree.  I think that's the way it was originally written, but I will change it.

File: features/step_definitions/admin_steps.rb [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+When(/^I enter the author "(.*?)"$/) do |author|
+  fill_in "Author", with: author
+end
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb:65]: Diito my comment on line 58

File: features/step_definitions/admin_steps.rb [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+Then(/^I see the book "(.*?)"$/) do |title|
+  expect(page).to have_content(title)
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb:68]: This is probably fine, but you could be more specific using either an xpath or a css selector

File: features/step_definitions/admin_steps.rb [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+Then(/^I see the book author "(.*?)"$/) do |author|
+  expect(page).to have_content(author)
+end
+
+Given(/^there is a book named "(.*?)"$/) do |title|
+  create(:book, title: title)
+end
+
+Given(/^the price of the book is "(.*?)"$/) do |price|
+  @book.price_cents = price
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb:88]: I assume this works? I would think you'd have to do `@book.update(price_cents: price)`

Also, you're given a dollar amount (eg $12.50). The value is in cents, so make sure you convert it. Or better yet, have a model method to do it for you.

File: features/step_definitions/admin_steps.rb [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+end
+
+When(/^I delete the Book with the Title "(.*?)"$/) do |title|
+  find('tr', text: title).click_link("Delete")
+end
+
+Then(/^I see a prompt requesting that I confirm my decision to delete the book$/) do
+  expect(page.driver.browser.switch_to.alert.text).to have_content("Are you sure you want to delete this?")
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb:96]: Good job! I know for me, this simple thing was one of the hardest things to do

File: features/step_definitions/admin_steps.rb [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+end
+
+When(/^I confirm my decision to delete the book$/) do
+  page.driver.browser.switch_to.alert.accept
+end
+
+Then(/^I don't see "(.*?)"$/) do |title|
+  expect(page).to_not have_content(title)
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb:104]: It looks like it's not needed right now, but if you end up creating a `I see "(.*?)"` step, combine the two like this:

```
Then(/^I( don't)? see "(.*?)"$/) do |negate, content|
  if negate
    expect(page).to_not have_content(content)
  else
    expect(page).to have_content(content)
  end
end
```

File: features/step_definitions/admin_steps.rb [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+end
+
+When(/^I edit the book with the title "(.*?)"$/) do |title|
+  find('tr', text: title).click_link("Edit")
+end
+
+Given(/^there is a book named "(.*?)" valued at "(.*?)"$/) do |arg1, arg2|
+  pending # express the regexp above with the code you wish you had
```
Comment (dkniffin) [inline features/step_definitions/admin_steps.rb]: Hmm...it looks like this isn't used. If so, remove it.

File: spec/factories/admin_users.rb [hunk 18]
```diff
@@ -0,0 +1,6 @@
+FactoryGirl.define do
+  factory :admin_user do
+
```
Comment (dkniffin) [inline spec/factories/admin_users.rb:3]: `admin_user` doesn't have any attributes?

File: spec/models/admin_user_spec.rb [hunk 19]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe AdminUser, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (dkniffin) [inline spec/models/admin_user_spec.rb:4]: either add some tests here or delete the file.

File: features/administration.feature (hunk 5 above)
Comment (dkniffin) [inline features/administration.feature:60]: I think the "Update Book" one is fine, but this comment was actually directed at all the changes in the file. I should have clarified.

File: features/step_definitions/admin_steps.rb (hunk 8 above)
Comment (dacur) [inline features/step_definitions/admin_steps.rb]: Yep, I took them out and the tests passed, thanks

File: features/step_definitions/admin_steps.rb (hunk 15 above)
Comment (dacur) [inline features/step_definitions/admin_steps.rb:96]: Thanks!  For me, too!

**Pull request merged**